A Blog post about Quarkus LangChain4j and OPA based guardrails (#2785)
quarkusio/quarkusio.github.io: _posts/2026-07-08-quarkus-langchain4j-opa-guardrails.adoc

---
layout: post
title: 'Guardrails with OPA Policies in Quarkus LangChain4j'
date: 2026-07-08
tags: quarkus LangChain4j OPA AI security guardrails LLM
synopsis: 'Implementing input guardrails with Open Policy Agent (OPA) for secure LLM applications.'
author: fabiobrz
---

== Introduction

As LLMs become integral to enterprise applications, securing the prompts they receive is critical.

https://docs.quarkiverse.io/quarkus-langchain4j/dev/index.html[Quarkus LangChain4j] is a Quarkus extension that allows
users to integrate AI into applications, and the

> https://quarkus.io/quarkus-workshop-langchain4j/[Quarkus LangChain4j Workshop] is a good way learn about and practice

Reviewer: ```suggestion
https://quarkus.io/quarkus-workshop-langchain4j/[Quarkus LangChain4j Workshop] is a good way to learn about and practice
```

---
layout: post
title: 'Guardrails with OPA Policies in Quarkus LangChain4j'
date: 2026-07-08
tags: quarkus LangChain4j OPA AI security guardrails LLM
synopsis: 'Implementing input guardrails with Open Policy Agent (OPA) for secure LLM applications.'
author: fabiobrz
---

== Introduction

As LLMs become integral to enterprise applications, securing the prompts they receive is critical.

https://docs.quarkiverse.io/quarkus-langchain4j/dev/index.html[Quarkus LangChain4j] is a Quarkus extension that allows
users to integrate AI into applications, and the

> https://quarkus.io/quarkus-workshop-langchain4j/[Quarkus LangChain4j Workshop] is a good way to learn and practice about

Reviewer: ```suggestion
https://quarkus.io/quarkus-workshop-langchain4j/[Quarkus LangChain4j Workshop] is a good way learn about and practice the fundamentals
```

the fundamentals.

Regarding security, https://quarkus.io/quarkus-workshop-langchain4j/section-1/step-09[Step 9] describes _input

> guardrails_, showcasing an example on how a separate LLM can be used to detect _prompt injection_ attacks.

Reviewer: ```suggestion
guardrails_, showcasing an example of how a separate LLM can be used to detect _prompt injection_ attacks.
```

However, this approach has costs, both in terms of tokens and performance.

> A deterministic pattern matching can catch obvious attacks faster, and delegating to the LLM becomes the last option,

Reviewer: ```suggestion
Deterministic pattern matching can catch obvious attacks faster, and delegating to the LLM becomes the last option,
```

		{"pattern": "i\\s+have\\s+special\\s+access", "weight": 0.85},
		{"pattern": "act\\s+as\\s+(if|though)", "weight": 0.8},
		{"pattern": "trust\\s+me\\s+and", "weight": 0.8},
		{"pattern": "(delete|remove|drop|cancel)\\s+all\\s+(the\\s+)?(bookings|reservations|data|records)", "weight": 0.85},
		{"pattern": "ignore\\s+(all\\s+)?(your|the|these)\\s+(rules|guidelines|policies)", "weight": 0.95},
		{"pattern": "you\\s+(must|should|have\\s+to)\\s+obey", "weight": 0.9},
		{"pattern": "speak\\s+(only\\s+)?in\\s+[a-z]+\\s+(from\\s+now|going\\s+forward)", "weight": 0.8},
		{"pattern": "(secret|hidden|special)\\s+(code|password|key|token)", "weight": 0.85},
	]

	exact_matches := [weight |
		p := exact_patterns[_]
		contains(text_lower, p.pattern)
		weight := p.weight
	]

	regex_matches := [weight |
		p := regex_patterns[_]
		regex.match(p.pattern, text_lower)
		weight := p.weight
	]

	all_matches := array.concat(exact_matches, regex_matches)
	score := max(array.concat(all_matches, [0.0]))
}
----

The above https://www.openpolicyagent.org/docs/latest/policy-language/[Rego] snippet defines an OPA policy that parses
the prompt to find either exact or regex matches of malicious content.

> A deterministic input guardrail could use such policy to detect a prompt injection attack, and the policy can be

Reviewer: ```suggestion
Deterministic input guardrail could use such policy to detect a prompt injection attack, and the policy can be
```

updated by the security team, independently of the application business logic and lifecycle.

> Now, how to quickly integrate OPA policies evaluation into our example Java application?

Reviewer: ```suggestion
Now, how to quickly integrate OPA policy evaluation into our example Java application?
```

== Adding Wasm to the recipe

Several source languages can be compiled to Wasm, and WebAssembly runtimes exist that allow integrating Wasm modules
into applications.

By adding https://github.com/StyraOSS/opa-java-wasm[Styra Open Policy Agent WebAssembly Java SDK] to the recipe, we
can obtain an application that is capable of fast and in-process OPA policy evaluation, e.g.:

[source,java]
----
var policy = OpaPolicy.builder().withPolicy(new File("policy.wasm")).build();
// ...
var result = policy.evaluate(input);
----

The OPA-based guardrail scans the prompt deterministically, but how can it fall back to the LLM when the pattern match
is inconclusive?

> The answer is represented by https://www.openpolicyagent.org/docs/latest/extensions/[OPA extensions], that the

Reviewer: The answer lies in OPA extensions, which the Styra SDK implements as custom builtins.

Please format the article correctly, I cannot make suggestions when things are cut in the middle of a sentence.

https://github.com/StyraOSS/opa-java-wasm[Styra Open Policy Agent WebAssembly Java SDK] implements as custom _builtins_,
i.e. Java functions that can be registered and called when needed.

Let's improve the Rego policy definition in order to catch suspicious words that could represent a weak signal of a
prompt injection attack, and the logic to delegate to the LLM evaluation in such a case:

[source,rego]
----

> package promptinjection

Reviewer: isn't it the same as above? It's long enough not to have it twice.

Author: It is not the same: there's additional logic to implement suspicious words matching that represents a weak signal of prompt injection and conditional escalation to the LLM. I am removing the redundant parts.

		{"word": "all instructions above", "weight": 0.5},
		{"word": "different assistant", "weight": 0.5},
		{"word": "tell me the", "weight": 0.3},
	]

	suspicious_matches := [weight |
		w := suspicious_words[_]
		contains(text_lower, w.word)
		weight := w.weight
	]

	all_matches := array.concat(array.concat(exact_matches, regex_matches), suspicious_matches)
	score := max(array.concat(all_matches, [0.0]))
}

injection_score := pattern_score if {
	pattern_score > 0.7
} else := llm_score(input.text) if {
	pattern_score > 0.0
	pattern_score <= 0.7
} else := 0.0

allow := result if {
	result := injection_score <= 0.7
}
----

The above Rego code snippet defines the logic to call a _builtin_ function, i.e. `llm_score`, when the
deterministic evaluation returns a score that is under the value signaling a potential prompt injection attack.

The `llm_score` function must be declared in a `capabilities.json` file that will be used when compiling Rego into
Wasm, for the OPA compiler to accept it:

[source,bash]
----
opa build -t wasm -e promptinjection/allow \
  --capabilities src/main/resources/policies/capabilities.json \
  src/main/resources/policies/prompt-injection.rego
----

> The guardrail is (and I'm not even sure about the italic here) the WebAssembly policy: when it needs deeper analysis,

Reviewer: Well, please remove my own comment.

Author: :man_facepalming: oh well, sorry, fixing.

it calls the LLM as a custom builtin:

....
User Input
    │
    ▼
┌─────────────────────────────────────────────────────┐
│            OPA Wasm Policy (342 KB)                 │
│                                                     │
│  1. Pattern matching (30 exact + 12 regex rules)    │
│     ├─ score > 0.7  ──►  BLOCKED     (< 1ms, $0)    │
│     ├─ score = 0.0  ──►  ALLOWED     (< 1ms, $0)    │
│     └─ 0 < score ≤ 0.7                              │
│          │                                          │
│  2.      └─► llm_score() custom builtin             │
│               │    OPA calls the LLM as a           │
│               │    data source, not a guardrail     │
│               ▼                                     │
│          OPA makes the final decision               │
└─────────────────────────────────────────────────────┘
    │
    ▼
  CustomerSupportAgent (Quarkus + LangChain4j AI Service)
....

Let's move on to the actual implementation.

== _Optional_ - Switch the original application backend to Ollama + Mistral

> The original application uses OpenAI services, and such configuration is actually a production grade one that works for

Reviewer: The original application uses OpenAI services, which also work for this example.

Reviewer: please avoid claim like production grade.

the current example, too.

Feel free to skip this additional step if you want to keep the original workshop example approach and use OpenAI APIs.

See the https://docs.ollama.com/quickstart[Ollama documentation quickstart] about installing Ollama and running an LLM,
we chose Mistral:

[source,shell]
----
ollama pull mistral
ollama run mistral
----

The sample application configuration should be updated accordingly with the required dependency:

[source,xml]
----
<dependency>
  <groupId>io.quarkiverse.langchain4j</groupId>
  <artifactId>quarkus-langchain4j-ollama</artifactId>
</dependency>
----

and configuration:

[source,properties]
----
quarkus.langchain4j.ollama.base-url=http://localhost:11434
quarkus.langchain4j.ollama.chat-model.model-id=mistral
quarkus.langchain4j.ollama.chat-model.log-requests=true
quarkus.langchain4j.ollama.chat-model.log-responses=true
quarkus.langchain4j.ollama.chat-model.temperature=0.0
quarkus.langchain4j.ollama.timeout=120s
----

See https://github.com/fabiobrz/quarkus-langchain4j-opa-guards[the companion sample application] for details.

> Time to code the new input guardrail, let's see how to do that.

Reviewer: ```suggestion
Time to code the new input guardrail. Let's see how to do that.
```

                return failure("Prompt injection detected by OPA policy");
            }
            return success();
        } catch (Exception e) {
            throw new RuntimeException("Failed to evaluate OPA policy", e);
        }
    }
}
----

As you can see, the class implements `InputGuardrail`, as the original example `PromptInjectionGuard` class does.
It loads a configured Wasm module which is obtained by compiling the OPA policy Rego definition into Wasm, and it uses
`com.styra.opa.wasm.OpaPolicy` to evaluate the input.

We'll add a new property to the application configuration, to externalize the location of the OPA Wasm policy file:

[source,properties]
----
# OPA Policy Configuration
# Path to the OPA Wasm policy file (relative to classpath resources)
opa.policy.path=/policies/prompt-injection.wasm
----

The interesting part, though, is that the `OPAPromptInjectionGuard` class also holds a reference to an instance of
`PromptInjectionDetectionService`, that is set by the constructor. This is the existing LLM-based detection service
from the workshop, reused as the builtin's implementation.

> Such component is used in the implementation of the _builtin_ `llm_score` function, which is registered by the

Reviewer: ```suggestion
This component is used in the implementation of the _builtin_ `llm_score` function, which is registered by the
```

`OpaPolicy.builder()` invocation.

Once we have the OPA guardrail, we'll have to update the existing `CustomerSupportAgent` class, and specifically the

> value of the `@InputGuardrail` annotation on the `chat()` method:

Reviewer: The snippet uses the plural form.

Author: Fixed, thanks

    2026-07-02 15:07:00,542 INFO  [dev.langchain4j.quarkus.workshop.OPAPromptInjectionGuard] (vert.x-worker-thread-1) OPA guardrail - BLOCKED: prompt injection detected
    2026-07-02 15:07:00,544 WARN  [dev.langchain4j.quarkus.workshop.CustomerSupportAgentWebSocket] (vert.x-worker-thread-1) Input guardrails detected a security issue: The guardrail dev.langchain4j.quarkus.workshop.OPAPromptInjectionGuard_ClientProxy failed with this message: Prompt injection detected by OPA policy
    Exception in CustomerSupportAgentWebSocket.java:26
              24      public String onTextMessage(String message) {
              25          try {
            → 26              return customerSupportAgent.chat(message);
              27          } catch (InputGuardrailException e) {
              28              Log.warnf(e, "Input guardrails detected a security issue: %s", e.getMessage());: dev.langchain4j.guardrail.InputGuardrailException: The guardrail dev.langchain4j.quarkus.workshop.OPAPromptInjectionGuard_ClientProxy failed with this message: Prompt injection detected by OPA policy
            at dev.langchain4j.guardrail.InputGuardrailExecutor.execute(InputGuardrailExecutor.java:68)

> The first two never touch the LLM. The third one shows the custom _builtin_ in action, when  the policy escalates on

Reviewer: ```suggestion
The first two never touch the LLM. The third one shows the custom _builtin_ in action, when the policy escalates on
```

[cols="1,3,3,3", options="header"]
|===
| | LLM-only guardrail | Wasm guardrail | Hybrid (Wasm delegating to LLM)

| *Latency*
| Varies by model and deployment, commonly seconds
| Sub-millisecond
| Varies by model and deployment, commonly seconds when falling back to LLM-based validation

| *Cost*
| Per-call token costs
| $0
| Per-call token costs when falling back to LLM-based validation

| *Determinism*
| Probabilistic -- same input can get different results
| Deterministic -- same input, same result, every time
| Probabilistic -- same input can get different results when falling back to LLM-based validation

| *Auditability*
| Black box
| Policy is a readable Rego file, version-controlled
| Mixed, black box for LLM-based validation and auditable for Wasm

| *Dependencies*
| Requires LLM API availability
| Self-contained binary, runs anywhere
| Requires LLM API availability when falling back to LLM-based validation
|===

== Conclusion

Implementing OPA-based guardrails in Quarkus LangChain4j provides:

* *Performance*: Sub-millisecond for pattern-matched prompts, vs seconds for LLM-based checks

> * *Cost*: LLM related costs decrease as the Wasm based validation success rate increases

Reviewer: ```suggestion
* *Cost*: LLM-related costs decrease as the Wasm-based validation success rate increases
```

* *Flexibility*: Easy policy updates without code changes
* *Separate policy life cycle management*: The lifecycle of OPA-based guardrails can be managed independently of the
application lifecycle.
* *Platform independent implementation*: A Wasm OPA guardrail works the same across different platforms
(JVM/Node/Python, etc.)

The combination of Quarkus LangChain4j's guardrail framework, OPA policies and Wasm integration creates a robust,

> production-ready security layer for AI applications.

Reviewer: "production-ready security layer" is a bit too strong here. We have no benchmark, no test coverage, and no deployment evidence yet.  Some for "production-grade guardrails"

Author: Removed
